src/ClientHandle.cpp

```diff
@@ -3126,6 +3126,35 @@ void cClientHandle::SendUseBed(const cEntity & a_Entity, int a_BlockX, int a_Blo
 
 
 
+void cClientHandle::SendUnlockRecipe(UInt32 a_RecipeId)
+{
+	m_Protocol->SendUnlockRecipe(a_RecipeId);
+}
+
+
+
+
+
+void cClientHandle::SendInitRecipes(UInt32 a_RecipeId)
+{
+	m_Protocol->SendInitRecipes(a_RecipeId);
+}
+
+
+
+
+
+void cClientHandle::HandleCraftRecipe(UInt32 a_RecipeId)
+{
+	auto SlotArea = m_Player->GetWindow()->GetSlotArea(0);
+	auto slotAreaCrafting = static_cast<cSlotAreaCrafting *>(SlotArea);
+	slotAreaCrafting->LoadRecipe(*m_Player, a_RecipeId);
```

Comment:
This makes a number of dangerous assumptions. You assume:
- `GetWindow()` doesn't return `nullptr`
- The window has a crafting grid
- That crafting grid is always the zeroth slot area

Use `GetWindowType()` to verify it's the window type you expect. Then instead of relying on slot areas ordering, I suggest putting a `LoadRecipe` method into the concrete window class:

```suggestion
	auto * Window = m_Player->GetWindow();
	if (Window == nullptr)
	{
		return;
	}
	
	auto WindowType = Window->GetWindowType();
	if (Window->GetWindowType() == cWindow::wtInventory)
	{
		static_cast<cInventoryWindow *>(Window)->LoadRecipe(*m_Player, a_RecipeId);
	}
	else if (Window->GetWindowType() == cWindow::wtWorkbench)
	{
		static_cast<cCraftingWindow *>(Window)->LoadRecipe(*m_Player, a_RecipeId);
	}
```

Reply:
Ok, I get your point. I think it is hard to send the package without having the window open (in the minecraft client), but better safe than sorry.

How to handle the implementation in `cInventoryWindow` and `cCraftingWindow`? Duplicate the code? Better idea?

Reply:
True, we don't expect the minecraft client to do this. But generally, we try to be fault tolerant when it comes to protocol inputs. You could have a non-standard client, or have edge cases where a Cuberite plugin breaks the vanilla client's assumptions.

The window's `LoadRecipe` just needs to call `cSlotAreaCrafting::LoadRecipe`.

Reply:
:+1: 

As you showed in your suggested changes it can be either `cInventoryWindow` or `cCraftingWindow`, both are only derived from `cWindow`. Both would need to have the `LoadRecipe` method. which is a two-liner, so doable.
Is there any better way, derived somehow from each other, ....? I would duplicate the two lines right now unless you have any other fancy idea.

Reply:
Yes, duplicating a two line function is okay.